### ISSUE TITLE
GPS/IMU is terrible.

### ISSUE THREAD

**Issue by superjax** [byu-magicc/roscopter#8]
Per my email discussion with @nonrevlb, the EKF is really strugging on simulated data.

Even after using a perfect IMU and perfect GPS, we get really bad performance.

![image](https://cloud.githubusercontent.com/assets/7344014/18445355/0a51443c-78d2-11e6-9b0d-b1d614eb92f4.png)

**Comment by superjax** [comment]
So I just played around with it for a while, and I'm still getting the poor performance.  @nonrevlb Do you think you could take a minute and go through my jacobians?  I have found the latex source code, and I added it to the repo.  I did actually find an error when I went through them just now, I but I would appreciate another set of eyes.

I have turned of the velocity and angle portion of the GPS update.  I also turned off all noise on the IMU and the GPS in the ros_copter_ekf_testing branch of byu-magicc/fcu_sim.  Go ahead and try it with the ekf_development branch.

**Comment by nonrevlb** [comment]
I haven't looked at the jacobian yet because I wanted to see how it was behaving first and I'm having trouble with the ros_copter_ekf_testing branch of byu-magicc/fcu_sim.  It is only publishing imu messages of (0,0,-9.8).  It seems that in gazebo_imu_plugin.cpp vdot is always zero.  

Since the imu z is always -9.8 it never triggers the flying_ bool to start running gps and imu updates and the ekf always just publishes an estimate of zero.

**Comment by superjax** [comment]
That's weird...  Okay.  So there was a comment in the original source code that says that sometimes, GetRelativeLinearAccel returns zero.  Which would be why you're getting vdot to always be zero.

We could do it the way that ETHZ did it, but lets add in the coriolis terms. - The following psuedocode should work

``` c
vdot = (v_current- v_prev) / dt;
omega = GetRelativeAngularVel();
coriols  = v_current.cross(omega);
total_accel = vdot + coriolis + gravity;
```

What do you think?

**Comment by nonrevlb** [comment]
I implemented vdot that way and it seems to be working pretty well.

I looked through the jacobian on the second page of that pdf.  I think I found a few errors.  Could you double check these?
- C07 should be -st*cs*u + sp*ct*cs*v + cp*ct***cs***w
- C13 should be **ct***ss
- C87 should be positive

I implemented these changes but it still doesn't work very well.  I'll keep looking into it but I was hoping you could check these 3 changes to make sure I'm not thinking about this wrong.

**Comment by superjax** [comment]
You're right.  Thanks for going through those.  I know that's a huge pain.

I think I may have figured out (at least partially) what is going on.  It has to do with the IMU plugin. Somehow, gravity is being applied backwards (I tested this by making the quad go up and down over and over again, and watching the simulated IMU data.). 

It should be 

``` c
  gravity_W_ = world_->GetPhysicsEngine()->GetGravity();
  ...
  math::Vector3 velocity_current_W = link_->GetWorldLinearVel();
  math::Vector3 velocity_current_B = link_->GetRelativeLinearVel();

  math::Vector3 vdot = (velocity_current_B - velocity_prev_W_) / dt;
  math::Vector3 angular_velocity_B = link_->GetRelativeAngularVel();
  math::Vector3 coriolis = angular_velocity_B.Cross(velocity_current_B);

  math::Vector3 gravity_B = C_W_I.RotateVector(gravity_W_);
  math::Vector3 total_acceleration_B = vdot + coriolis - gravity_B;
```

Notice, the final line.  Instead of the previous version, where we were doing

```
total_acceleration_B = vdot + coriolis + gravity_B;
```

we are flipping gravity

```
total_acceleration_B = vdot + coriolis - gravity_B;
```

I am in the middle of some work on `fcu_sim` right now, so it'll be a huge pain to merge that change myself.  I was wondering if you could try putting that in, and see what happens.

My next test, to make sure that IMU is working properly would be to send the quad back and forth in the x-direction such that it reaches steady-state velocity, observe IMU to make sure it responds as expected, and again in the y-direction.  Do you think you could do that, and make sure that works?  You could also see it if in the EKF if the roll and pitch estimates are terrible.

I'm sorry this is such a huge pain. It's supposed to be easy.  I guarantee though, that this will be helpful to everyone else in the lab.

**Comment by superjax** [comment]
By the way, I am working on an altitude estimator right now, and these changes fixed the problem on my side.  I am only doing altitude right now, so I am not going to be able to see angle changes.

**Comment by nonrevlb** [comment]
Switching the gravity didn't fix it.  I think the problem may come from the angle estimates so I've been looking at the components of the orientation quaternion.  I'm seeing something kind of weird that I don't understand.  With the gravity switched to being subtracted, I get:
![minus_gravity](https://cloud.githubusercontent.com/assets/13068053/18682894/4044e42c-7f2b-11e6-9fd7-ea91eb545476.png)
It goes the opposite direction from truth.

When I switch back to adding gravity, the bumps happen the same but it slowly diverges:
![plus_gravity](https://cloud.githubusercontent.com/assets/13068053/18682938/6bb6ebdc-7f2b-11e6-9f2d-4368fa829a17.png)

Do you know what might cause this?

**Comment by nonrevlb** [comment]
On equation 8 of the EKF Jacobian paper, I believe that the x acceleration should be **-**g*st.  This effects equation 9 and the imu updated step.  Changing this in the code make the x position estimate stop going the opposite direction.

Here is a recent graph of the estimate I get with the GPS helping keep it on track:
![improved_ekf](https://cloud.githubusercontent.com/assets/13068053/18686326/c8ff93c2-7f38-11e6-8ea1-8197bac7f1f9.png)

The IMU part of the estimate still tends to overshoot the actual values.  Is this normal?  I expected it to be more accurate when we are using perfect IMU messages.

Also, would you like me to make the changes I've pointed out to the PDF and push that?

**Comment by nonrevlb** [comment]
Actually, the y estimate was also going the wrong way until I switched the sign of the y acceleration in equation 8 (now it is **+**g*ct*sp).

**Comment by superjax** [comment]
Haha!!  Nice.  That is awesome.  Fantastic work.  Kalman Filter can be so frustrating, way to stick with it.

The reason that the IMU doesn't work perfectly is because there is simulated drag.  The drag introduces acceleration in the x and y dimensions which is not taken into account in our dynamic model (we could take it into account, you can talk to David about this if you're interested, the drag dynamics are actually observable given position data).  This additional acceleration makes our attitude estimate wrong.  The transient response is very well estimated by the gyros, but the prolonged attitude estimation tends towards level.  That is, unless we start estimating drag.

Okay, so - here's what you should do.  Take the branch with the updated jacobians, and create a pull request, then assign me to it. That's the official way that we will merge all your changes into master.  You have the ability to merge it straight in yourself, but it's sort of poor form.  Do the same thing with the simulator.   The pull request lets us talk about the changes, and I can make suggestions, and eventually we'll merge it in and it will be the official master.

**Comment by superjax** [comment]
This is fixed by #9.
